frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/traverse/TraverseSyncToParent.java

```diff
@@ -0,0 +1,133 @@
+import com.virtuslab.gitmachete.frontend.resourcebundles.GitMacheteBundle;
+import com.virtuslab.gitmachete.frontend.ui.api.table.BaseEnhancedGraphTable;
+import com.virtuslab.gitmachete.frontend.vfsutils.GitVfsUtils;
+
+@ExtensionMethod({GitMacheteBundle.class, GitVfsUtils.class})
+public class TraverseSyncToParent {
+
+  private final Project project;
+  private final GitRepository gitRepository;
+  private final BaseEnhancedGraphTable graphTable;
+  private final IGitMacheteRepositorySnapshot repositorySnapshot;
+  private final INonRootManagedBranchSnapshot gitMacheteBranch;
+  private final @UI Runnable traverseNextEntry;
+
+  public TraverseSyncToParent(GitRepository gitRepository,
+      BaseEnhancedGraphTable graphTable,
+      IGitMacheteRepositorySnapshot repositorySnapshot,
+      INonRootManagedBranchSnapshot gitMacheteBranch,
+      @UI Runnable traverseNextEntry) {
+    this.project = gitRepository.getProject();
+    this.gitRepository = gitRepository;
+    this.graphTable = graphTable;
+    this.repositorySnapshot = repositorySnapshot;
+    this.gitMacheteBranch = gitMacheteBranch;
+    this.traverseNextEntry = traverseNextEntry;
+  }
+
+  @UIEffect
+  public void sync() {
+    @UI Runnable syncToRemoteRunnable = new TraverseSyncToRemote(gitRepository, graphTable, gitMacheteBranch,
+        traverseNextEntry)::sync;
+
+    switch (gitMacheteBranch.getSyncToParentStatus()) {
+      case MergedToParent :
+        if (!handleMergedToParent()) {
+          return;
+        }
+        break;
+
+      case InSyncButForkPointOff :
+      case OutOfSync :
+        if (!handleOutOfSyncOrInSyncButForkPointOff(syncToRemoteRunnable)) {
+          return;
+        }
+        break;
+
+      default :
+        break;
+    }
+
+    graphTable.queueRepositoryUpdateAndModelRefresh(syncToRemoteRunnable);
+  }
+
+  @UIEffect
+  private boolean handleMergedToParent() {
+    val branchLayout = repositorySnapshot.getBranchLayout();
+    val currentBranchIfManaged = repositorySnapshot.getCurrentBranchIfManaged();
+    val slideOutDialog = MessageDialogBuilder.yesNoCancel(
+        getString("action.GitMachete.BaseTraverseAction.dialog.merged-to-parent.title"),
+        getString(
+            "action.GitMachete.BaseTraverseAction.dialog.merged-to-parent.text.HTML").fmt(
+                gitMacheteBranch.getName()))
+        .cancelText(getString("action.GitMachete.BaseTraverseAction.dialog.cancel-traverse"));
+
+    switch (slideOutDialog.show(project)) {
+      case MessageConstants.YES :
+        new SlideOutBackgroundable(getString("action.GitMachete.BaseSlideOutAction.task.title"),
+            gitMacheteBranch, gitRepository, currentBranchIfManaged, branchLayout, graphTable) {
+          @Override
+          public void onSuccess() {
+            graphTable.queueRepositoryUpdateAndModelRefresh(traverseNextEntry);
+          }
+        }.queue();
+        return false;
```

Comment:
Huh that's tricky ;D so it's now down to the newly-created backgroundable and NOT down to `sync()` to continue the traversal 🤔

Reply:
I'd add a comment in code about this for clarity

Reply:
:yes: